### PR TITLE
Update to Ghidra 10.1.3

### PR DESCRIPTION
**Pull request by mpi3d**
After spending some time updating this Flatpak, I've managed to make it work with the latest Ghidra version.

This should solve #6, #12, #14, #15, #16, and also NationalSecurityAgency/ghidra#3726.

I changed many things: For exemple Ghidra is installed in `/app/ghidra`. I removed `ghidra.properties` options. Also I use `cp` instead of `install`. Fell free to revert those changes. I've just cleaned all those things to make it work in a first place. But now that it works...

### CODE REVIEW

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json:99]: I can take bets that this isn't going to build. There's no network support inside the flatpak-builder, the dependencies need to be fetched ahead of time.

**Comment by hadess** [comment]
> But now that it works...

I can hazard a guess and say that it won't build.

If you have tested the patch locally though, I would be grateful if you could file a PR for it upstream so it gets in the next point release, which we could then pick up.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [inline org.ghidra_sre.Ghidra.json:65]: @hadess This is solving the network issue

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:99]: Line `65` is solving the network issue

**Comment by mpi3d** [comment]
> > But now that it works...
> 
> I can hazard a guess and say that it won't build.
> 
> If you have tested the patch locally though, I would be grateful if you could file a PR for it upstream so it gets in the next point release, which we could then pick up.

I've tested that locally. And it works. Really.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:65]: Wait... Ok I get it. We can't change the build options: `Can't specify build-args in sandboxed build`

**Comment by mpi3d** [comment]
I'll find another way to resolve this network issue.

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [reply org.ghidra_sre.Ghidra.json:65]: This isn't something I can merge...

**Comment by hadess** [comment]
> I'll find another way to resolve this network issue.

Ideally gradle support would be added to the flatpak-builder-tools:
https://github.com/flatpak/flatpak-builder-tools/issues/37
So that tarballs could be prefetched.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:65]: I'll try to use another method.

**Comment by mpi3d** [comment]
I did found a more pleasant way to download dependencies using Flatpak Builder and then build with `gradle` offline.
I also improved the patch

**Comment by mpi3d** [comment]
Sounds like the build works fine!
@hadess Is it ok for merging?

**Comment by hadess** [comment]
It will need reviewing. As a quick review, given the number of changes, the changes would need to be split up in as many commits as it makes sense, first adding the json files with the dependencies (and explaining how they were generated), updating the build system, adding the missing releases, etc. all in small, trivial, commits so that the review is easier.

Thanks for working on that!

**Comment by mpi3d** [comment]
Ok. I'll try to make that.

**Comment by mpi3d** [comment]
I split this in multiples commits. Just ping me if you don't understand something.
Concerning the deps, for Ghidra (f3e4088ab26929ba48bc4a1be8d53f609429f007), I just followed the [Ghidra Dev Guide](https://github.com/NationalSecurityAgency/ghidra/blob/master/DevGuide.md#manual-download-instructions) and also [`fetchDependencies.gradle`](https://github.com/NationalSecurityAgency/ghidra/blob/883f5a687a3bc2b1b2425dffd2ac7724df1218cb/gradle/support/fetchDependencies.gradle#L54-L148).
For Gradle (3639edce785ebbc2232c7c0b444755ea3474d1e6) it was more hard. I had to run Gradle offline and add deps for each network error concerning dependences downloads. It sounds hacky but I don't think that Gradle support listing the deps for a project because it just resolve them at compile time.

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json]: Why is this "improved"? Doesn't seem to be any different, and I prefer longer options as that means I don't need to look up the meaning of argument.

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json]: This also seems to be the exact same line that got removed. Maybe this whole commit can be dropped?

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json:81]: Your commit subject is far too sparse for this change. You need to explain the impact of the change in the commit subject (I'm guessing it's "Always use upstream launch.properties"?), and explain how that change was implemented in the commit message.

File: TODO
Comment (hadess) [inline TODO:5]: What allows you to remove the TODO item? Is it fixed by the previous commit? If so, you might as well merge those 2 commits, so it's clear when the problem was fixed.

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json:46]: Explain why this is a necessary change in the commit message.

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json:60]: Why use `cp` when `install` is a one-liner?

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json:62]: Ditto.

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json:64]: Ditto.

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json:55]: You'll probably want to reorder the commits (`git rebase -i origin/master`) so this commit still builds. The dependencies haven't been added to the manifest, so it's not going to work.

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [inline org.ghidra_sre.Ghidra.json:132]: As mentioned in a previous review comment, you need to explain how you ended up with this file, so that it can eventually be regenerated and updated.

File: last_modified.patch
Comment (hadess) [inline last_modified.patch:9]: Can you please file a PR against Ghidra and reference it in the commit message?

**Review by hadess — CHANGES_REQUESTED** [review]
Quite a few nits to fix before this can be merged, but apart from "how to generate the gradle deps", it's just git surgery. Good job.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json]: Right. I didn't thought that longer argument could help for a better understanding. I'll correct that.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json]: It can. Order doesn't matter here.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:81]: Ok

File: TODO
Comment (mpi3d) [reply TODO:5]: Yeah, this should be fixed by commit 5ef862d1a0c79d27dbb6c7099e8e4bf3a5f58ef3.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:46]: Ok

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:60]: I never used `install` before. But I guess it's time to start learning how to use it.

File: last_modified.patch
Comment (mpi3d) [reply last_modified.patch:9]: You mean to create a pull request on Ghidra's git ([NationalSecurityAgency/ghidra](https://github.com/NationalSecurityAgency/ghidra))? I don't think they will accept. Trying anyway.

Comment (hadess) [reply last_modified.patch:9]: Even if your PR isn't accepted, at least we can use it to link to solutions that were merged, or an explanation as to why it's still a downstream patch, something which would be lost in the review in this PR.

Comment (mpi3d) [reply last_modified.patch:9]: Ok

Comment (mpi3d) [reply last_modified.patch:9]: Pull request: NationalSecurityAgency/ghidra#4107

**Comment by mpi3d** [comment]
Forgot to mention NationalSecurityAgency/ghidra#4107

**Comment by mpi3d** [comment]
@hadess This has been a long time. What's happening? Do I need to do something else?

**Comment by hadess** [comment]
> @hadess This has been a long time. What's happening? Do I need to do something else?

I've been waiting for your updates on my review comments since the beginning of April, in particular, this one:
https://github.com/flathub/org.ghidra_sre.Ghidra/pull/18#discussion_r836394197
which I can't merge your changes without, because it would then be impossible for me to update the dependencies.

Once you've reviewed all the comments, and as you fix them, please make sure to mark the threads as "resolved" along with a comment explaining how you fixed it (eg. whether you modified an existing commit, dropped a patch, or something else).

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json]: Corrected

File: TODO
Comment (mpi3d) [reply TODO:5]: Merged

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:46]: Done: 4002738b9491dac35ea1bda4d7ddd5859c206d03

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json]: Done

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:54]: Done

File: last_modified.patch
Comment (mpi3d) [reply last_modified.patch:9]: Done: 56fdbe1f76d8890f7ec60deb44fe0b9f8edecad9

**Comment by mpi3d** [comment]
@hadess Oh, ok. I did forget to mark them as resolved. Sorry, I'm not an expert in Git.

I did resolve them all.
https://github.com/flathub/org.ghidra_sre.Ghidra/pull/18#discussion_r836394197 is solved too.

Also, you should know that I will still be there after this merge. So in case of any future problem with dependencies I can help. You just need to contact me.

Tell me if there is something else to do.

**Comment by hadess** [comment]
I'll look at this on Monday, but I'm still not happy with the way `gradle-dependencies.json` is generated. We need a better way to do  this.

**Comment by mpi3d** [comment]
I don't think there is any better way to generate this. No one is providing such script:
+ https://github.com/flathub/com.tonikelope.MegaBasterd/blob/master/maven-dependencies.yaml
+ https://github.com/flathub/es.estoes.wallpaperDownloader/blob/master/maven-dependencies.yaml

File: org.ghidra_sre.Ghidra.json
Comment (hadess) [reply org.ghidra_sre.Ghidra.json:86]: I don't understand how that solves the 2x scaling, now that we're not copying the launch.properties file for users to modify. The update TODO reference would be https://github.com/bkerler/ghidra_installer/blob/master/install-ghidra.sh#L24-L31

**Comment by hadess** [comment]
> I don't think there is any better way to generate this. No one is providing such script:
> 
>     * https://github.com/flathub/com.tonikelope.MegaBasterd/blob/master/maven-dependencies.yaml
> 
>     * https://github.com/flathub/es.estoes.wallpaperDownloader/blob/master/maven-dependencies.yaml

I made some notes at https://github.com/flatpak/flatpak-builder-tools/issues/37#issuecomment-1121368390

**Comment by hadess** [comment]
Can you please check whether this works? https://github.com/flatpak/flatpak-builder-tools/pull/276

You might need to declare the `.exe` downloads separately.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:86]: It's not solving the 2x scaling. It's only solving the use of a symbolic link in `/bin`.
So indeed we should replace the link by your.

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:51]: It's not required. I'll drop this commit

**Comment by mpi3d** [comment]
> Can you please check whether this works? [flatpak/flatpak-builder-tools#276](https://github.com/flatpak/flatpak-builder-tools/pull/276)
> 
> You might need to declare the `.exe` downloads separately.

I'm on it

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:46]: Done

File: org.ghidra_sre.Ghidra.json
Comment (mpi3d) [reply org.ghidra_sre.Ghidra.json:81]: Done

**Comment by mpi3d** [comment]
@hadess Sounds like the build failed because of an internal error.

**Comment by mpi3d** [comment]
bot, build

**Comment by mpi3d** [comment]
Better

**Comment by mpi3d** [comment]
bot, build

**Comment by mpi3d** [comment]
I hope this is ok for the merge.

**Comment by hadess** [comment]
I've reworked the commit messages and contents to my liking. I'll wait for the last 10.1.3 build to finish and will run a quick test. If all that passes, I'll push all this. Thanks very much for your help in getting all this updated.

**Comment by mpi3d** [comment]
Thanks

**Pull request merged**
